### PR TITLE
Model saving/restoring using EFF

### PR DESCRIPTION
**Pull request by tkornuta-nvidia**
- [x] EFF dependency in requirements.txt (0.1.0 release)
 - [x] Modified `ModelPT.save_to()`
 - [x] Modified `ModelPT.restore_from()`
 - [x] `ModelPT.__set_eff_save()`/`Model.__use_eff_save()` static methods for manual enabling/disabling saving using EFF (flag set to `True` by default).
 - [x] unit test - polished test adding artifact
 - [x] unit test - testing checkpoint encryption when using _eff_save_to()/_eff_restore_from() with provided encryption key.

`ModelPT.save_to()` logic:
 - if EFF present and __use_eff_save flag set: save .nemo using EFF
 - else: save .nemo using the old method

`ModelPT.restore_from()` logic:
 - if EFF present: try to restore the model using EFF, if fails: try to restore using the old method
 - else: restore .nemo using the old method

In short:
 - Restore will be able to load both the existing .nemo files and the new EFF-based .nemo files
 - All newly saved .nemo files will have EFF under the hood.

Note that from the content of the archive perspective, the new .nemo archive is a superset of the old format, i.e. the old .nemo file content:
```
 + model_config.yaml
 + model_weights.chkt
 + … (additional files a.k.a. artifacts)
```
The new .nemo file content:
```
 + manifest.yaml (<- this file is the only difference!)
 + model_config.yaml
 + model_weights.chkt
 + … (additional files a.k.a. artifacts)
```

### CODE REVIEW

File: nemo/core/classes/modelPT.py [hunk 1]
```diff
@@ -193,8 +197,74 @@ def save_to(self, save_path: str):
...
+            model_wights.chpt - model checkpoint
+
+        Args:
+            save_path: Path to .nemo file where model instance should be saved
+        """
+        # Add nemo rank check as well
+        if not is_global_rank_zero():
+            return
+
+        if _EFF_PRESENT_:
```
Comment (titu1994) [inline nemo/core/classes/modelPT.py:259]: Such a global flag is not a good design - what if a user has eff installed but what's nemo checkpoints (when not working with TLT workflow). Create a global variable, create a setter method to enable / disable eff type file saving, and use that flag here.

File: nemo/core/classes/modelPT.py [hunk 2]
```diff
@@ -261,6 +328,126 @@ def restore_from(
...
+                assert isinstance(model, nemo.collections.asr.models.EncDecCTCModel)
+                ```
+
+        Returns:
+            An instance of type cls
+        """
+        if not path.exists(restore_path):
+            raise FileNotFoundError(f"Can't find {restore_path}")
+
+        if _EFF_PRESENT_:
```
Comment (titu1994) [inline nemo/core/classes/modelPT.py:440]: Same comment as above.

File: nemo/core/classes/modelPT.py (hunk 1 above)
Comment (tkornuta-nvidia) [reply nemo/core/classes/modelPT.py:259]: "what if a user has eff installed but what's nemo checkpoints" - not sure what you mean by that, can you please explain?

Comment (tkornuta-nvidia) [reply nemo/core/classes/modelPT.py:259]: > Create a global variable, create a setter method to enable / disable eff type file saving, and use that flag here.

Not sure what you mean by "setter" method... who and when should call it?

Can you please take a look at `save_to` and `restore_from` and tell me what is missing?

Comment (tkornuta-nvidia) [reply nemo/core/classes/modelPT.py:259]: Or more generally: what do you think on the logic of `save_to` and `restore_from` provided in the description? What is missing?

Comment (tkornuta-nvidia) [reply nemo/core/classes/modelPT.py]: Ok, we discussed that with @titu1994, and the comment was not about bad "global flag" design (which is commonly used in NeMo for handling other conditional dependencies), but rather the lack of option for programmatically enabling/disabling using EFF.

I honestly see low usability of this, but added additional _MODEL_EFF_SAVE flag with setter/getter static methods to model, so the user, in his code, will be able to manually turn off saving .nemo files with EFF.

Comment (titu1994) [reply nemo/core/classes/modelPT.py]: Yep, poor wording on my part. Thanks for his Tom

Comment (tkornuta-nvidia) [reply nemo/core/classes/modelPT.py]: The opposite - thanks for reviewing and discussion.

File: requirements/requirements.txt [hunk 3]
```diff
@@ -10,3 +10,4 @@ scikit-learn
 omegaconf>=2.0.2
 hydra-core>=1.0.3
 transformers>=3.1.0
+nvidia-eff @ http://sqrl.nvidia.com/dldata/pip-scratch/nvidia-eff/nvidia_eff-0.1.0-py3-none-any.whl
```
Comment (blisc) [inline requirements/requirements.txt:13]: Please remove this from requirements. Running `pip install -r requirement.txt` results in a loop
```
Collecting nvidia-eff@ http://sqrl.nvidia.com/dldata/pip-scratch/nvidia-eff/nvidia_eff-0.1.0-py3-none-any.whl
  WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPConnection object at 0x7f4be827b8d0>: Failed to establish a new connection: [Errno -2] Name or service not known',)': /dldata/pip-scratch/nvidia-eff/nvidia_eff-0.1.0-py3-none-any.whl
  WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPConnection object at 0x7f4be827bcc0>: Failed to establish a new connection: [Errno -2] Name or service not known',)': /dldata/pip-scratch/nvidia-eff/nvidia_eff-0.1.0-py3-none-any.whl
```

**Review by blisc — CHANGES_REQUESTED** [review]

File: nemo/core/classes/modelPT.py [hunk 4]
```diff
@@ -35,11 +35,20 @@
...
+except ImportError:
+    _EFF_PRESENT_ = False
+
+
 __all__ = ['ModelPT']
 
 _MODEL_CONFIG_YAML = "model_config.yaml"
 _MODEL_WEIGHTS = "model_weights.ckpt"
 _MODEL_IS_RESTORED = False
+_MODEL_EFF_SAVE = True
```
Comment (titu1994) [inline nemo/core/classes/modelPT.py:51]: By default this should be False.

File: nemo/core/classes/modelPT.py [hunk 5]
```diff
@@ -1019,3 +1226,13 @@ def __is_model_being_restored() -> bool:
     def __set_model_restore_state(is_being_restored: bool):
         global _MODEL_IS_RESTORED
         _MODEL_IS_RESTORED = is_being_restored
+
+    @staticmethod
+    def __set_eff_save(use_eff_save: bool):
```
Comment (titu1994) [inline nemo/core/classes/modelPT.py:1231]: Do not use __ here, it makes them name mangled and users cannot use them from ModelPT or it's subclasses.

File: nemo/core/classes/modelPT.py [hunk 6]
```diff
@@ -1019,3 +1226,13 @@ def __is_model_being_restored() -> bool:
...
+        global _MODEL_EFF_SAVE
+        _MODEL_EFF_SAVE = use_eff_save
+
+    @staticmethod
+    def __use_eff_save() -> bool:
```
Comment (titu1994) [inline nemo/core/classes/modelPT.py:1236]: Same here, remove the leading __

**Review by titu1994 — CHANGES_REQUESTED** [review]

File: requirements/requirements.txt (hunk 3 above)
Comment (tkornuta-nvidia) [reply requirements/requirements.txt:13]: hmmm, ok, this is bad.

I was using `pip install -e .`, also tested `reinstall.sh`, worked properly.

Let me try to replicate that in a new environment.

File: nemo/core/classes/modelPT.py (hunk 4 above)
Comment (tkornuta-nvidia) [reply nemo/core/classes/modelPT.py:51]: Leaving as is for now, we will discuss this.

File: nemo/core/classes/modelPT.py (hunk 5 above)
Comment (tkornuta-nvidia) [reply nemo/core/classes/modelPT.py:1231]: wasn't sure how/where you want to use it, but ok, you are right, let's make it "public"

File: nemo/core/classes/modelPT.py (hunk 6 above)
Comment (tkornuta-nvidia) [reply nemo/core/classes/modelPT.py:1236]: wasn't sure how/where you want to use it, but ok, you are right, let's make it "public"

File: requirements/requirements.txt (hunk 3 above)
Comment (blisc) [reply requirements/requirements.txt:13]: I don't think this file is accessible outside the internal network, so please retest when not on the company VPN

Comment (tkornuta-nvidia) [reply requirements/requirements.txt:13]: Hmmmm, not sure about that, there are plenty of tools (including DALI) that are distributed this way.
@DEKHTIARJonathan can you please comment on this? 

Independently, let me test that "outside" of NVIDIA network...

Comment (tkornuta-nvidia) [reply requirements/requirements.txt:13]: @blisc you are right! I just tried to reinstall NeMo on my laptop (outside of VPN) and got the same error as you:
```
Collecting nvidia-eff@ http://sqrl.nvidia.com/dldata/pip-scratch/nvidia-eff/nvidia_eff-0.1.0-py3-none-any.whl
  WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPConnection object at 0x7ff14a589050>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known')': /dldata/pip-scratch/nvidia-eff/nvidia_eff-0.1.0-py3-none-any.whl
```
Ok, this is bad...

**Review by titu1994 — APPROVED** [review]
Aligned. Let's look into the pip install issue (or remove temporarily) and update requirements.txt once its available.

File: requirements/requirements.txt (hunk 3 above)
Comment (okuchaiev) [reply requirements/requirements.txt:13]: yes, please, let's remove from reqs until generally available

File: nemo/core/classes/modelPT.py (hunk 5 above)
Comment (tkornuta-nvidia) [reply nemo/core/classes/modelPT.py]: done

File: nemo/core/classes/modelPT.py (hunk 6 above)
Comment (tkornuta-nvidia) [reply nemo/core/classes/modelPT.py]: done

File: requirements/requirements.txt (hunk 3 above)
Comment (tkornuta-nvidia) [reply requirements/requirements.txt]: ok, removed the dependency, thanks @blisc for catching that...! 👍
